### PR TITLE
Cleans up project warnings 📱

### PR DESCRIPTION
**Pull request by ashfurrow**
Eigen's gotten into a very poor state, with compiler warnings and static analyzer warnings everywhere. With all this noise, valid warnings get lost and we lose advantages of using a compiler. 

This PR addresses the warnings to get us here:

screen shot 2017-05-18 at 11 40 12 am

It does this in a few ways:

- Isolate deprecated UIKit APIs into Objective-C files and silence the warnings.
- Address all Swift/UIKit/Foundation deprecation warnings.
- Clean up static analyzer warnings.
- Updates to latest Xcode recommended settings.

This is a WIP because it requires extensive testing and I don't want to merge before the eigen release is out.

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          2 Errors
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">No CHANGELOG changes made</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Developer Specific file shouldn’t be changed</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Needs testing on a Phone if change is non-trivial</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_travis">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by ashfurrow** [comment]
@erikdstock please do ask about everything! These should all be well-documented. If they're not, let me know.

**Comment by ashfurrow** [comment]
I wonder if this would be a good time to enable the "treat warnings as errors" option 🤔 Maybe not locally but on CI alone somehow?

File: Artsy/View_Controllers/Embedded/Generics/AREmbeddedModelsViewController.m [hunk 1]
```diff
@@ -356,8 +356,8 @@ - (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView
         return view;
     }
 
-    NSAssert(YES, @"Should not be able to get here");
```
Comment (alloy) [inline Artsy/View_Controllers/Embedded/Generics/AREmbeddedModelsViewController.m]: 😂

Comment (alloy) [inline Artsy/View_Controllers/Embedded/Generics/AREmbeddedModelsViewController.m]: Very positive assert

**Comment by alloy** [comment]
Ah yeah, great stuff, @ashfurrow 🙌 

LGTM 👍

**Comment by ashfurrow** [comment]
It would be pretty easy to pass `-Werror` only during `make test` so that we could treat warnings as errors on CI, but locally wouldn't impact development. I think it's worth trying for a while, any opinions?

**Comment by orta** [comment]
I think that's reasonable - it's not like they're hidden from you on day to day and then just appear 30 minutes later on CI.

File: Artsy.xcodeproj/xcshareddata/xcschemes/Artsy.xcscheme [hunk 2]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <Scheme
-   LastUpgradeVersion = "0700"
+   LastUpgradeVersion = "0830"
    version = "1.3">
    <BuildAction
       parallelizeBuildables = "YES"
```
Comment (orta) [inline Artsy.xcodeproj/xcshareddata/xcschemes/Artsy.xcscheme]: These are the changes that Danger is failing the build for. So this PR is fine to merge if red - so long as someone has done a quick check for the test results

Comment (ashfurrow) [inline Artsy.xcodeproj/xcshareddata/xcschemes/Artsy.xcscheme]: Yeah yeah, I'm working on a merge now. Shouldn't have changed the PR titled until I had pushed.

**Comment by ashfurrow** [comment]
Still getting deprecation warnings from a pod we're importing. Going to look into later.

**Comment by ashfurrow** [comment]
Okay, I've got this wrapped up. Expecting Danger to be 🔴 but otherwise CI should pass.

**Comment by ashfurrow** [comment]
Okay, this is passing on CI and is ready for final review /cc @orta.

File: Artsy/View_Controllers/Embedded/Generics/AREmbeddedModelsViewController.m [hunk 3]
```diff
@@ -356,8 +356,8 @@ - (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView
...
-    return nil;
+    NSAssert(NO, @"Should not be able to get here");
```
Comment (orta) [inline Artsy/View_Controllers/Embedded/Generics/AREmbeddedModelsViewController.m]: In theory this assert now does nothing, but it acts as a comment - so I'll let it fly

**Comment by orta** [comment]
interesting - @ashfurrow this _should_ fail on my new build

screen shot 2017-05-24 at 15 46 39

**Comment by orta** [comment]
:O - the flag didn't work?

screen shot 2017-05-24 at 15 48 25

**Comment by ashfurrow** [comment]
The warning pictured is from a pod, but it is odd. I'll make sure it fails with the make command before merging 👍

**Comment by ashfurrow** [comment]
Just a typo, it's `OTHER_CFLAGS` with an underscore. I'll push a fix.

**Comment by ashfurrow** [comment]
Okay, getting similar warnings when I enable the flag as on CI:

```
❌  /Users/distiller/eigen/Artsy/View_Controllers/Fair/Guides/ARFairGuideViewController.m:3:9: 'ARSwitchView.h' file not found

#import "ARSwitchView.h"
        ^


▸ Compiling ARAuctionArtworkTableViewCell.m
** BUILD FAILED **
```

Going to investigate.

File: Makefile [hunk 4]
```diff
@@ -6,6 +6,7 @@ APP_PLIST = Artsy/App_Resources/Artsy-Info.plist
 STICKER_PLIST = Artsy\ Stickers/Info.plist
 PLIST_BUDDY = /usr/libexec/PlistBuddy
 DEVICE_HOST = platform='iOS Simulator',OS='9.0',name='iPhone 6'
+OTHER_CFLAGS = OTHER_CFLAGS="\$$(inherited) -Werror"
```
Comment (ashfurrow) [inline Makefile]: Okay so the trick is we need `$(inherited)` to get passed as a flag or else the ones specified in the project aren't included. This line has a kind of double-escaping, `$$` gets escaped to `$` by `make` and then`\$` escapes it from the shell so it won't try and run some command called `inherited`.

**Comment by ashfurrow** [comment]
I've tested this locally that it fails when I have an unused variable, so I think this is set to go after the CI passes 👍

**Comment by orta** [comment]
Rock

**Comment by alloy** [comment]
Awesome! Thanks so much 💯

**Pull request merged**